storage/postgresql/save.go

```diff
@@ -60,3 +43,47 @@ func (b *PostgresBackend) AfterSave(evt *nostr.Event) {
       ORDER BY created_at DESC OFFSET 100 LIMIT 1
     )`, evt.PubKey, evt.Kind)
 }
+
+func deleteBeforeSaveSql(evt *nostr.Event) (string, []any, bool) {
+	// react to different kinds of events
+	var (
+		query        = ""
+		params       []any
+		shouldDelete bool
+	)
+	if evt.Kind == nostr.KindSetMetadata || evt.Kind == nostr.KindContactList || (10000 <= evt.Kind && evt.Kind < 20000) {
+		// delete past events from this user
+		query = `DELETE FROM event WHERE pubkey = $1 AND kind = $2`
+		params = []any{evt.PubKey, evt.Kind}
+		shouldDelete = true
+	} else if evt.Kind == nostr.KindRecommendServer {
+		// delete past recommend_server events equal to this one
+		query = `DELETE FROM event WHERE pubkey = $1 AND kind = $2 AND content = $3`
+		params = []any{evt.PubKey, evt.Kind, evt.Content}
+		shouldDelete = true
+	} else if evt.Kind >= 30000 && evt.Kind < 40000 {
+		// NIP-33
+		d := evt.Tags.GetFirst([]string{"d"})
+		if d != nil {
+			query = `DELETE FROM event WHERE pubkey = $1 AND kind = $2 AND tagvalues && ARRAY[$3]`
+			params = []any{evt.PubKey, evt.Kind, d.Value()}
+			shouldDelete = true
+		}
+	}
+
+	return query, params, shouldDelete
+}
+
+func saveEventSql(evt *nostr.Event) (string, []any, error) {
+	const query = `INSERT INTO event (
+	id, pubkey, created_at, kind, tags, content, sig)
+	VALUES ($1, $2, $3, $4, $5, $6, $7)
+	ON CONFLICT (id) DO NOTHING`
+
+	var (
+		tagsj, _ = json.Marshal(evt.Tags)
```

Comment:
There's opportunity for error handling here but I didn't want to change any of the existing behavior without double checking.